### PR TITLE
Enable attribute `deprecated_in_future`

### PR DESCRIPTION
**Pull request by tcharding**
This is a bit curly because of current `serde` logic, fixed in patch 4. All the other patches are trivial.

### CODE REVIEW

**Comment by apoelstra** [comment]
Well, the offender here is the `SerdeHash::from_slice_delegated` method right? Currently it passes through to `Self::from_slice` which is deprecated.

Instead we could "manually" implement this by calling `<&[u8; $len]>::try_from` and then converting that error to a `FromSliceError`. Then if that passes we can call `Self::from_byte_array`.

Arguably we don't even need to construct the `FromSliceError` because in our in-crate usage, we just throw this error away and call `E::invalid_length`. But this is a method in a public API that returns a `FromSliceError` so I guess we ought to do it properly.

File: hashes/src/sha256t.rs [hunk 1]
```diff
@@ -71,7 +71,6 @@ where
     pub fn from_engine(e: HashEngine) -> Hash<T> { from_engine(e) }
 
     /// Copies a byte slice into a hash object.
-    #[deprecated(since = "TBD", note = "Use `from_byte_array` instead.")]
```
Comment (Kixunil) [inline hashes/src/sha256t.rs:74]: We should deprecate the other uses instead of un-deprecating this one.

**Review by Kixunil — COMMENTED** [review]
Concept ACK except for removing deprecations.

**Comment by Kixunil** [comment]
@apoelstra I'm not really happy with `SerdeHash`. It seems to be just exposing some internal guts and it doesn't look to me like that should be even required.

**Comment by apoelstra** [comment]
Also -- win for the lint. Frustrating that this is off by default and super obscure, because this is exactly the kind of thing we'd like to notice **before** publishing deprecations.

@Kixunil I agree. I think this may have been a "code reuse" trait that wound up being public because of our own inter-crate visibility requirements. But it's not obvious to me that it even actually needs to *be* a trait.

**Comment by tcharding** [comment]
Includes change to error message in test because we still test against error message strings in places.

**Comment by apoelstra** [comment]
Needs rebase

**Comment by tcharding** [comment]
Refactored slightly the `deserialize` impl while fixing my rebase mistake.

**Comment by apoelstra** [comment]
I am going to ACK this because it lets us enable the deprecation flag. I'm not thrilled with the serde stuff (though I think it's better than the old version, since we're just exposing a couple of opaque structs, and not a whole trait). I don't have a good solution for it, and I think it's out of scope for this PR.

**Review by apoelstra — DISMISSED** [review]
ACK f8e6321113ded9efdd2436de88c98b6d228d9ef4 successfully ran local tests

**Comment by tcharding** [comment]
Rebase had conflicts with other attribute we just added, I made all the crates the same (except `addresses` which is missing `#![doc(test(attr(warn(unused))))]`).

**Review by apoelstra — APPROVED** [review]
ACK f5e56856c64f729c61f496d6cc22d4740ae639c7 successfully ran local tests

File: hashes/src/sha256t.rs (hunk 1 above)
Comment (tcharding) [reply hashes/src/sha256t.rs]: FTR, this is done.

**Comment by tcharding** [comment]
Rebased and fixed typo in commit brief messsage `s/Fully deprecated/Fully deprecate/`. No other changes.

**Review by apoelstra — DISMISSED** [review]
ACK ac023de6f63c0b411823b619659bebd12feafebc successfully ran local tests

**Comment by tcharding** [comment]
Rebased and remove `OutPoint` stuff because that went in already when we moved `OutPoint` to `primitives` - I think, CI will tell me if not.

**Review by apoelstra — APPROVED** [review]
ACK a78e81289bb55ac540b9ccbb0c8adc1629aeee5f successfully ran local tests

**Comment by tcharding** [comment]
And same question, do we want this in the `hashes` release?

**Comment by apoelstra** [comment]
Also yes -- I'll one-ack merge this since it's been sitting so long.

**Comment by apoelstra** [comment]
Needs rebase now.

**Comment by tcharding** [comment]
Needs #3455

**Comment by tcharding** [comment]
Rebased to pick up unrelated fix. No changes.

**Review by apoelstra — APPROVED** [review]
ACK e68da281ff26fea093810a0a277b15adb62e061a successfully ran local tests

**Pull request merged**